### PR TITLE
HAProxy TLS Termination

### PR DESCRIPTION
**Pull request by wyattrees**
WIP

Asks user for an SSL certificate and key to pass to the HAProxy charm

### CODE REVIEW

File: anvil-python/anvil/commands/haproxy.py [hunk 1]
```diff
@@ -59,10 +100,35 @@ def __init__(
...
+            key = key_file.read()
+        if not os.path.isdir(HAPROXY_CERTS_DIR):
+            mkdir(HAPROXY_CERTS_DIR)
+        with open(os.path.join(HAPROXY_CERTS_DIR, "haproxy.pem"), "w") as combined_file:
+            combined_file.write(key + cert)
+        
+        self.variables["charm_haproxy_config"]["services"] = self._TLS_SERVICES_CONFIG
+        LOG.debug(f"HAProxy prompt variables: {self.variables}")
+
+    def extra_tfvars(self) -> dict:
```
Comment (SK1Y101) [inline anvil-python/anvil/commands/haproxy.py:129]: should probably be dict[str, Any]

File: anvil-python/anvil/commands/haproxy.py [hunk 2]
```diff
@@ -59,10 +100,35 @@ def __init__(
             "Deploying HAProxy",
             refresh,
         )
+        self.variables = {"charm_haproxy_config": {}}
```
Comment (SK1Y101) [inline anvil-python/anvil/commands/haproxy.py:103]: might want to make this `self.variables: dict[str, Any] = {...`

File: anvil-python/anvil/commands/haproxy.py [hunk 3]
```diff
@@ -34,11 +38,48 @@
 HAPROXY_UNIT_TIMEOUT = (
     1200  # 15 minutes, adding / removing units can take a long time
 )
+HAPROXY_CERTS_DIR = os.path.join(environ['HOME'], "certs/")
+LOG = logging.getLogger(__name__)
+
+
+def validate_cert_file(filepath: str) -> None:
```
Comment (SK1Y101) [inline anvil-python/anvil/commands/haproxy.py:45]: this seemed unhappy when the path was None, might be worth adding a step ensuring this is a string before using it as a filepath

File: anvil-python/anvil/commands/haproxy.py [hunk 4]
```diff
@@ -34,11 +38,48 @@
...
+        if "BEGIN PRIVATE KEY" not in f.read():
+            raise ValueError("Invalid key file")
 
 
 class DeployHAProxyApplicationStep(DeployMachineApplicationStep):
     """Deploy HAProxy application using Terraform"""
 
+    _HAPROXY_QUESTIONS = {
+        "ssl_cert": questions.PromptQuestion(
+            "Path to SSL Certificate for HAProxy: ",
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:65]: This prompt is ending with `: ` where as the `ssl_key` one does not include those characters. Could you please update the one that needs to be updated? :slightly_smiling_face:

File: anvil-python/anvil/commands/haproxy.py [hunk 5]
```diff
@@ -34,11 +38,48 @@
...
+        "ssl_key": questions.PromptQuestion(
+            "Path to private key for the SSL certificate",
+            validation_function=validate_key_file,
+        ),
+    }
+    _TLS_SERVICES_CONFIG = f"""- service_name: haproxy_service
+  service_host: "0.0.0.0"
+  service_port: 443
+  service_options:
+    - balance leastconn
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:77]: What is the non-related TLS configuration if we do not setup TLS? How can we guarantee that those options do not afffect the HAProxy setup?

**Review by skatsaounis — CHANGES_REQUESTED** [review]
I added some inline comments for you to take a look. Overall, the PR looks good. Thank you @wyattrees :slightly_smiling_face: .

One last need: We need to also add a README file that describes how we can enable TLS termination with Anvil. Could you please prepare a doc like this and add it to your PR? Thank you.

File: anvil-python/anvil/commands/haproxy.py (hunk 3 above)
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py]: Made None a valid option (in the case where user wants to proceed without TLS)

File: anvil-python/anvil/commands/haproxy.py (hunk 5 above)
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py:82]: The only changes to the `services.yaml` for TLS termination are the `service_port` and `crts` keys. If the user opts to skip TLS configuration, the charm's default `services.yaml` is used

File: README.md [hunk 6]
```diff
@@ -34,6 +34,11 @@ ubuntu@infra1:~$ maas-anvil cluster bootstrap \
 Note: You will be asked for a `virtual_ip` during installation of the HAProxy charm, if `accept-defaults` is omitted.
 Pass an empty value to disable it, or any valid IP to enable; the Keepalived charm will be installed to enable connecting to HA MAAS using the VIP.
 
+#### TLS Termination at HA Proxy
+While deploying HA Proxy, MAAS Anvil will ask you for filepaths pointing to an SSL certificate and private key. If passed, HA Proxy will be configured to use the given certificate and key for TLS termination. To skip TLS configuration, enter nothing when prompted for the certificate and key files (this is the behavior if `--accept-defaults` is passed).
```
Comment (skatsaounis) [inline README.md:38]:
```suggestion

While deploying HAProxy, MAAS Anvil will ask you for filepaths pointing to an SSL certificate and private key. If passed, HAProxy will be configured to use the given certificate and key for TLS termination. To skip TLS configuration, enter nothing when prompted for the certificate and key files (this is the behavior if `--accept-defaults` is passed).
```

File: anvil-python/anvil/commands/haproxy.py [hunk 7]
```diff
@@ -92,6 +133,7 @@ def __init__(
         )
         self.preseed = deployment_preseed or {}
         self.accept_defaults = accept_defaults
+        self.variables: dict[str, Any] = {"charm_haproxy_config": {}}
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:136]: I think that this is not needed now

```suggestion
```

File: anvil-python/anvil/commands/haproxy.py [hunk 8]
```diff
@@ -100,34 +142,55 @@ def has_prompts(self) -> bool:
...
-        variables = questions.load_answers(
-            self.client, self._KEEPALIVED_CONFIG
-        )
+        variables = questions.load_answers(self.client, self._HAPROXY_CONFIG)
         variables.setdefault("virtual_ip", "")
+        variables.setdefault("charm_haproxy_config", {})
 
         # Set defaults
         self.preseed.setdefault("virtual_ip", "")
+        self.preseed.setdefault("charm_haproxy_config", {})
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:151]: `charm_haproxy_config` is a Terraform variable that we are keeping empty to give our users the options to always override what we set with our variables. It is provided as the last resort to skip all the sane defaults and whatever logic we have with single variables. e.g. `virtual_ip`. So I would like you to introduce dedicated variables for questions `ssl_cert`, `ssl_key`, similar Terraform variables for them and in the end, produce the services yaml with Terraform.

So here we will have something like:

```suggestion
        self.preseed.setdefault("ssl_cert", "")
        self.preseed.setdefault("ssl_key", "")
```

File: anvil-python/anvil/commands/haproxy.py [hunk 9]
```diff
@@ -100,34 +142,55 @@ def has_prompts(self) -> bool:
...
             console=console,
             preseed=self.preseed.get("haproxy"),
             previous_answers=variables,
             accept_defaults=self.accept_defaults,
         )
 
-        variables["virtual_ip"] = keepalived_config_bank.virtual_ip.ask()
+        variables["virtual_ip"] = haproxy_config_bank.virtual_ip.ask()
+
+        cert_filepath = haproxy_config_bank.ssl_cert.ask()
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:163]: let's move this to Terraform. I like the idea better.

File: anvil-python/anvil/commands/haproxy.py [hunk 10]
```diff
@@ -100,34 +142,55 @@ def has_prompts(self) -> bool:
...
+            if not os.path.isdir(HAPROXY_CERTS_DIR):
+                mkdir(HAPROXY_CERTS_DIR)
+            with open(
+                os.path.join(HAPROXY_CERTS_DIR, "haproxy.pem"), "w"
+            ) as combined_file:
+                combined_file.write(key + cert)
+            variables["charm_haproxy_config"]["services"] = (
+                self._TLS_SERVICES_CONFIG
+            )
+            variables["haproxy_port"] = 443
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:179]: Let's keep only this line but set it in `extra_tfvars` below. It is something that is not coming by prompt but it is set based on prompt answers

File: anvil-python/anvil/commands/haproxy.py [hunk 11]
```diff
@@ -12,10 +12,11 @@
 # implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:14]:
```suggestion
# limitations under the License.

```

**Review by skatsaounis — CHANGES_REQUESTED** [review]
Thank you for updating the PR. Here is hopefully the last round of suggestions.

File: README.md [hunk 12]
```diff
@@ -31,6 +31,11 @@ ubuntu@infra1:~$ maas-anvil cluster bootstrap \
     --accept-defaults
 ```
 
+#### TLS Termination at HA Proxy
+While deploying HA Proxy, MAAS Anvil will ask you for filepaths pointing to an SSL certificate and private key. If passed, HA Proxy will be configured to use the given certificate and key for TLS termination. To skip TLS configuration, enter nothing when prompted for the certificate and key files.
```
[the remark this answers is not in the document]
Comment (wyattrees) [reply README.md]: updated

File: anvil-python/anvil/commands/haproxy.py (hunk 7 above)
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py]: removed

File: anvil-python/anvil/commands/haproxy.py (hunk 8 above)
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py]: moved to a different terraform variable, which merges with `charm_haproxy_config` (takes precedent)

File: anvil-python/anvil/commands/haproxy.py (hunk 9 above)
Comment (wyattrees) [reply anvil-python/anvil/commands/haproxy.py:174]: omitting since we aren't actually passing the file contents as a string, but just a file path with both the cert and key concatenated

File: anvil-python/anvil/commands/haproxy.py [hunk 13]
```diff
@@ -100,34 +149,60 @@ def has_prompts(self) -> bool:
...
+                cert = cert_file.read()
+            with open(key_filepath) as key_file:
+                key = key_file.read()
+            if not os.path.isdir(HAPROXY_CERTS_DIR):
+                mkdir(HAPROXY_CERTS_DIR)
+            with open(
+                os.path.join(HAPROXY_CERTS_DIR, "haproxy.pem"), "w"
+            ) as combined_file:
+                combined_file.write(key + cert)
+            variables["haproxy_services_yaml"] = self._TLS_SERVICES_CONFIG
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:187]: This line should be deleted from here and added only to extra_tfvars. We do not want to save it to the database as a user answer.

File: anvil-python/anvil/commands/haproxy.py [hunk 14]
```diff
@@ -100,34 +149,60 @@ def has_prompts(self) -> bool:
...
-        )
+        questions.write_answers(self.client, self._HAPROXY_CONFIG, variables)
 
     def extra_tfvars(self) -> dict[str, Any]:
         variables: dict[str, Any] = questions.load_answers(
-            self.client, self._KEEPALIVED_CONFIG
+            self.client, self._HAPROXY_CONFIG
         )
-        variables["haproxy_port"] = 80
+        if self.use_tls_termination:
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:201]: Please add `variables["haproxy_services_yaml"] = self._TLS_SERVICES_CONFIG` here, after loading user answers from the variables

Comment (skatsaounis) [reply anvil-python/anvil/commands/haproxy.py:201]: In addition we could also remove ssl_cert, ssl_key from variables dict since Terraform does not need them

**Review by skatsaounis — CHANGES_REQUESTED** [review]

**Comment by wyattrees** [comment]
@skatsaounis this is ready for review once again. https://github.com/canonical/maas-charms/pull/194 still needs to be merged & released, but you can test with the `latest/edge/wyatt-test` channel for the `maas-region` charm in the meantime

**Review by skatsaounis — APPROVED** [review]
LGTM +1

**Comment by skatsaounis** [comment]
Enabler for #44
